src/main/java/com/znsio/teswiz/runner/DeviceSetup.java

```diff
@@ -82,17 +83,68 @@ static ArrayList<String> setupAndroidExecution() {
     static void verifyAppExistsAtMentionedPath() {
         String appPath = Setup.getFromConfigs(APP_PATH);
         LOGGER.info(String.format("Update path to Apk: %s", appPath));
-        if(appPath.equals(NOT_SET)) {
+        if (appPath.equals(NOT_SET)) {
             appPath = getAppPathFromCapabilities();
+            appPath = convertAppPathToFilePathIfNeeded(appPath);
             Setup.addToConfigs(APP_PATH, appPath);
-            String capabilitiesFileName = Setup.getFromConfigs(CAPS);
-            checkIfAppExistsAtTheMentionedPath(appPath, capabilitiesFileName);
         } else {
+            appPath = convertAppPathToFilePathIfNeeded(appPath);
             LOGGER.info(String.format("\tUsing AppPath provided as environment variable -  %s",
-                                      appPath));
+                    appPath));
         }
     }
 
+    private static String convertAppPathToFilePathIfNeeded(String appPath){
+        if (isAppPathAUrl(appPath)) {
+            String directoryPath = System.getProperty("user.dir") + File.separator + "temp" + File.separator + "sampleApps";
+            String fileName = appPath.split(File.separator)[appPath.split(File.separator).length - 1];
+            String filePath = directoryPath + File.separator + fileName;
+            createDirectoryIfNotExist(directoryPath);
+            downloadFileIfNotExist(appPath, filePath);
+            LOGGER.info("Changing value of appPath from URL to file path");
+            LOGGER.info("Before change, appPath value: "+ appPath);
+            appPath = filePath;
+            LOGGER.info("After change, appPath value: "+ filePath);
+        }
+        try{
+            new FileInputStream(appPath);
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException("File not found: " + appPath, e);
+        }
+        return appPath;
+    }
+    private static void createDirectoryIfNotExist(String directoryPath){
```

Comment:
The if check for directory exists should be done before calling this method.

That way we can eliminate if else.

Reply:
instead of checking before calling createDirectoryIfNotExist(), made the changes inside the method to make code look cleaner